lib/SVF-FE/LLVMModule.cpp

```diff
@@ -227,18 +228,79 @@ void LLVMModuleSet::initialize()
     }
 }
 
+std::vector<llvm::Function *> LLVMModuleSet::getLLVMGlobalFunctions(
+        llvm::GlobalVariable *global)
+{
+    class LLVMGlobalFunction {
+    // llvm.global_ctors and llvm.global_dtors
+    public:
+        unsigned priority;
+        llvm::Function *func;
+        LLVMGlobalFunction(){};
+        LLVMGlobalFunction(unsigned _priority, llvm::Function *_func)
+              : priority(_priority), func(_func){};
+        bool operator>(const LLVMGlobalFunction &other) const
+        {
+            return priority > other.priority;
+        }
+    };
+
+    std::priority_queue<LLVMGlobalFunction, std::vector<LLVMGlobalFunction>,
+                        greater<LLVMGlobalFunction>>
+            queue;
+    std::vector<llvm::Function *> result;
+
+    llvm::ConstantArray *globalFuncArray =
+            llvm::dyn_cast<llvm::ConstantArray>(global->getInitializer());
+    if (globalFuncArray) {
+        for (unsigned int i = 0; i < globalFuncArray->getNumOperands(); ++i) {
+            llvm::ConstantStruct *globalFuncItem =
+                    llvm::dyn_cast<llvm::ConstantStruct>(
+                            globalFuncArray->getOperand(i));
+            if (globalFuncItem) {
+                ConstantInt *priority = llvm::dyn_cast<llvm::ConstantInt>(
+                        globalFuncItem->getOperand(0));
+                Function *func = llvm::dyn_cast<llvm::Function>(
+                        globalFuncItem->getOperand(1));
+
+                if (priority && func) {
+                    queue.push(LLVMGlobalFunction((unsigned)priority
+                                                          ->getZExtValue(),
+                                                  func));
+                }
+            }
+        }
+    }
+
+    while (!queue.empty()) {
+        result.push_back(queue.top().func);
+        queue.pop();
+    }
+    return result;
+}
+
 void LLVMModuleSet::addSVFMain()
 {
     std::vector<Function *> init_funcs;
+    std::vector<Function *> destroy_funcs;
     Function * orgMain = 0;
     Module* mainMod = nullptr;
-    for (Module& mod : modules)
-    {
-        for (auto &func: mod)
-        {
-            if(func.getName().startswith(SVF_GLOBAL_SUB_I_XXX))
-                init_funcs.push_back(&func);
-            if(func.getName().equals(SVF_MAIN_FUNC_NAME))
+    for (Module &mod : modules) {
+        // Collect ctor and dtor functions
+        for (Module::global_iterator it = mod.global_begin(),
+                                     eit = mod.global_end();
+             it != eit; ++it) {
+            GlobalVariable *global = &*it;
```

Comment:
can we add a const qualifier for GlobalVariable and the below lines if we don't change their attributes?